### ISSUE TITLE
Add schema support for pyarrow map / struct / dictionary type 

### ISSUE THREAD

**Issue by raghavdixit99** [lancedb/lance#2341]
lancedb API throws `TypeError: Converting Pydantic type to Arrow Type: unsupported type typing.Dict[str, str].  ( same for Dict[int,str] )` . Issue is in lance not supporting the pyarrow map types.

To Reproduce :

```python
from lancedb.pydantic import Vector, LanceModel, Dict

class Content(LanceModel):
    id: int
    vector: Vector(1)
    metadata: Dict[int, str]

db = lancedb.connect("./lancedb")
tbl = db.create_table("test", schema=Content.to_arrow_schema())
```

**Comment by Zalasyu** [comment]
Yeah the problem if you store metadata as a string then when you call the invoke method with langchain after you converted your lancedb instance into a VectorStoreRetriever than you get an error and in the traceback it complains that metadata is of type str when get_relevent_documents expected a dictionary.

**Comment by ziyadperumpalli** [comment]
class SaveMail(LanceModel):
    id: str
    from_: Optional[Dict] = None
    to: Optional[List[Dict]] = None
    subject: str
    body: str
    full_text: str = embedder.SourceField()
    vector: Vector(embedder.ndims()) = embedder.VectorField()

table = db.create_table("memory", schema=SaveMail, exist_ok=True)

error :
    TypeError: Converting Pydantic type to Arrow Type: unsupported type typing.Dict.

Does anyone know how to fix this?

**Comment by westonpace** [comment]
I agree we should get better support here but it's not entirely clear what `Dict` should map to.

`Dict[X, Y]` can map to Arrow's map type.

`Dict[int, ???]` can probably map to Arrow's dictionary type but would it get matched by the above check and handled as a map type?

`Dict` without type hints is more challenging.  Perhaps we could map it to some kind of JSON type?  Or maybe wait until we have support for Variant?  Or I suppose we could use Arrow's union type.

**Comment by westonpace** [comment]
The 2.2 version of the file format has added support for Arrow's map type.  This does not solve this issue but helps to enable a solution.
